packages/expect/src/index.js

```diff
@@ -172,20 +172,18 @@ const makeRejectMatcher = (
     );
   }
 
-  let result;
-  try {
-    result = await actual;
-  } catch (e) {
-    return makeThrowingMatcher(matcher, isNot, e).apply(null, args);
-  }
-
-  throw new JestAssertionError(
-    utils.matcherHint(matcherStatement, 'received', '') +
-      '\n\n' +
-      `Expected ${utils.RECEIVED_COLOR('received')} Promise to reject, ` +
-      'instead it resolved to value\n' +
-      `  ${utils.printReceived(result)}`,
-  );
+  return actual
+    .then(v => {
+      const err = new JestAssertionError(
+        utils.matcherHint(matcherStatement, 'received', '') +
+          '\n\n' +
+          `Expected ${utils.RECEIVED_COLOR('received')} Promise to reject, ` +
+          'instead it resolved to value\n' +
+          `  ${utils.printReceived(v)}`,
+      );
+      return Promise.reject(err);
+    })
+    .catch(e => makeThrowingMatcher(matcher, isNot, e).apply(null, args));
```

Comment:
This `catch` has to be first

Reply:
Wait no, that won't work either. You have to use `.then(resolvedHandler, rejectedHandler)` to avoid handling the rejection here.

So just move your `catch` into the second argument of the `then`.

Reply:
https://repl.it/repls/RevolvingEvergreenBlackbuck
Is there any difference between A and B, besides syntax preference? Both a and b pass the test cases.

Reply:
In this case, yes. since we _expect_ it to fail we need to have our `catch` first so we can resolve the promise with it. But if it does _not_ reject we need to reject it ourselves.

If we add a `then` after the `catch`, it's not called. If we add our `catch` onto a `then`, we locally catch the rejection we want to bubble up